// @po-ui/ng-components/esm2015/lib/components/po-table/services/po-table.service.js
import { debounceTime } from 'rxjs/operators';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { fromEvent } from 'rxjs';
import { isTypeof } from '../../../utils/util';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
export class PoTableService {
    constructor(http) {
        this.http = http;
        this.headers = new HttpHeaders({
            'X-PO-No-Message': 'true'
        });
    }
    getFilteredItems(filteredParams) {
        const params = this.validateParams(filteredParams);
        return this.http.get(this.url, { headers: this.headers, params });
    }
    setUrl(url) {
        this.url = url;
    }
    scrollListener(componentListner) {
        return fromEvent(componentListner, 'scroll').pipe(debounceTime(100));
    }
    validateParams(params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    }
}
PoTableService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoTableService_Factory() { return new PoTableService(i0.ɵɵinject(i1.HttpClient)); }, token: PoTableService, providedIn: "root" });
PoTableService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
PoTableService.ctorParameters = () => [
    { type: HttpClient }
];
