// @po-ui/ng-components/esm2015/lib/components/po-chart/directives/po-resize-observer.directive.js
/// <reference types="resize-observer-browser" />
import { Directive, ElementRef, EventEmitter, Output } from '@angular/core';
import { Subscription, Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
export class PoResizeObserverDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.resize = new EventEmitter();
        this.subscription = new Subscription();
        this.chartWidthResize$ = new Subject();
    }
    ngOnDestroy() {
        if (this.isResizeObserverSupported) {
            this.observer.unobserve(this.elementRef.nativeElement);
            this.subscription.unsubscribe();
        }
    }
    ngOnInit() {
        if (this.isResizeObserverSupported) {
            this.observer = new window.ResizeObserver(() => {
                this.chartWidthResize$.next();
            });
            this.observer.observe(this.elementRef.nativeElement);
            this.subscription.add(this.chartWidthResize$.pipe(debounceTime(20)).subscribe(_ => {
                this.resize.emit();
            }));
        }
    }
    get isResizeObserverSupported() {
        return typeof window.ResizeObserver === 'function';
    }
}
PoResizeObserverDirective.decorators = [
    { type: Directive, args: [{
                selector: '[p-resize-observer]'
            },] }
];
PoResizeObserverDirective.ctorParameters = () => [
    { type: ElementRef }
];
PoResizeObserverDirective.propDecorators = {
    resize: [{ type: Output, args: ['p-resize-observer',] }]
};
