// @po-ui/ng-components/esm2015/lib/services/po-color/po-color.service.js
import { Injectable } from '@angular/core';
import { PoColorPaletteEnum } from '../../enums/po-color-palette.enum';
import { PoDefaultColors } from './po-colors.constant';
import * as i0 from "@angular/core";
const poColorPalette = Object.values(PoColorPaletteEnum);
export class PoColorService {
    constructor() {
        this.defaultColors = [];
    }
    /**
     * Avalia a propriedade `color` na lista de items passada. Caso sim, trata se é decimal ou string `po-color`. Caso não haja, retorna a cor default.
     *
     * @param data
     */
    getColors(data) {
        this.verifyIfHasColorProperty(data);
        return data.map((dataItem, index) => {
            if (dataItem.color) {
                dataItem.color = this.verifyIfIsPoColorPalette(dataItem.color);
                return dataItem;
            }
            const color = this.defaultColors[index];
            return Object.assign(Object.assign({}, dataItem), { color });
        });
    }
    verifyIfHasColorProperty(data) {
        const hasColorProperty = data.every(dataItem => { var _a; return dataItem.hasOwnProperty('color') && ((_a = dataItem.color) === null || _a === void 0 ? void 0 : _a.length) > 0; });
        if (!hasColorProperty) {
            this.defaultColors = this.getDefaultColors(data.length);
        }
    }
    verifyIfIsPoColorPalette(color) {
        if (poColorPalette.includes(color)) {
            return `po-${color}`;
        }
        return color;
    }
    getDefaultColors(length) {
        if (length === 1) {
            return PoDefaultColors[0];
        }
        const colorsLength = PoDefaultColors.length;
        if (length > colorsLength) {
            const quantityDuplicates = length / colorsLength;
            let colors = PoDefaultColors[colorsLength - 1];
            for (let i = 0; i <= quantityDuplicates; i++) {
                colors = colors.concat(PoDefaultColors[colorsLength]);
            }
            return colors;
        }
        return PoDefaultColors[length - 1];
    }
}
PoColorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoColorService_Factory() { return new PoColorService(); }, token: PoColorService, providedIn: "root" });
PoColorService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
