// @po-ui/ng-components/esm2015/lib/components/po-dynamic/po-dynamic.util.js
export function getGridColumnsClasses(smColumns, mdColumns, lgColumns, xlColumns, gridColumns, pullColumns) {
    const systemGrid = {
        sm: smColumns || gridColumns || 12,
        md: mdColumns || gridColumns || 6,
        lg: lgColumns || gridColumns || 4,
        xl: xlColumns || gridColumns || 3,
        pullSm: (pullColumns === null || pullColumns === void 0 ? void 0 : pullColumns.smPull) || 0,
        pullMd: (pullColumns === null || pullColumns === void 0 ? void 0 : pullColumns.mdPull) || 0,
        pullLg: (pullColumns === null || pullColumns === void 0 ? void 0 : pullColumns.lgPull) || 0,
        pullXl: (pullColumns === null || pullColumns === void 0 ? void 0 : pullColumns.xlPull) || 0
    };
    return `po-sm-${systemGrid.sm} po-pull-sm-${systemGrid.pullSm} po-md-${systemGrid.md} po-pull-md-${systemGrid.pullMd} po-lg-${systemGrid.lg} po-pull-lg-${systemGrid.pullLg} po-xl-${systemGrid.xl} po-pull-xl-${systemGrid.pullXl}`;
}
export function isVisibleField(field) {
    const containsVisible = 'visible' in field;
    return containsVisible ? field.visible : true;
}
