// @po-ui/ng-components/esm2015/lib/components/po-menu/services/po-menu-global.service.js
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class PoMenuGlobalService {
    constructor() {
        this.applicationMenu = new Subject();
        this.menus = new Subject();
        this.removedApplicationMenu = new Subject();
        // eslint-disable-next-line @typescript-eslint/member-ordering
        this.receiveApplicationMenu$ = this.applicationMenu.asObservable();
        // eslint-disable-next-line @typescript-eslint/member-ordering
        this.receiveMenus$ = this.menus.asObservable();
        // eslint-disable-next-line @typescript-eslint/member-ordering
        this.receiveRemovedApplicationMenu$ = this.removedApplicationMenu.asObservable();
    }
    sendApplicationMenu(menu) {
        this.applicationMenu.next(menu);
    }
    sendMenus(menus) {
        this.menus.next(menus);
    }
    sendRemovedApplicationMenu(id) {
        this.removedApplicationMenu.next(id);
    }
}
PoMenuGlobalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PoMenuGlobalService_Factory() { return new PoMenuGlobalService(); }, token: PoMenuGlobalService, providedIn: "root" });
PoMenuGlobalService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
